### DISCUSSION TITLE
[Recent Error of FileMesh or FileMeshGenerator] ABAQUS input file cannot be imported once elset setting.

### DISCUSSION THREAD

**Discussion by Ttw0626** [idaholab/moose#26001] [Q&A Modules: General] [answered]
Hi, moose-developers,

**introduction of problem**

Once I updated (fetch and rebase to master repo and updated libmesh) on 8.Nov, 
From my view, it should be error from new repo of libmesh.
the same file (abaqus input mesh file) cannot be read anymore.

the error output will be shown in below.
![ksnip_20231108-215322](https://github.com/idaholab/moose/assets/90584327/71e503a9-3995-4fad-ad6b-b87420d6f7d5)

`I attempted some different formats to figure out the locus of error`

**build-in abaqus input file test at /moose/test/tests/mesh/abaqus_input/testcube_elem_id_in.inp**
> **line 32** "10,"

`PASS`

**the same file with modification** 

> **line 31** *Elset, elset=_Part-1-1_top_S3
> **line 32** "10," ==>  "1, 2, 3, 4"

`ERROR`

**the same file with modification** 

> **line 31** *Elset, elset=_Part-1-1_top_S3
> **line 32** "10," ==>  "1,"

`ERROR`

**the same file with modification** 

> **line 31** *Elset, elset=_Part-1-1_top_S3
> **line 32** "10," ==>  "5,"

`PASS`

**the same file with modification** 

> **line 31** *Elset, elset=_Part-1-1_top_S3
> **line 32** "10," ==>  "12,"

`PASS`

**the same file with modification** 

>  **add two line before line 31**
>   *Elset, elset=_Part-1-1_top_S1, generate
>    "1, 10, 1"

`ERROR`

**thinking**

- we only can define the element set (subdomain in MOOSE) which should be uitlized as surface (sidesets in MOOSE) ?
- which part was modified within `FileMeshGenerator` to cause this different result compared to few days ago ?
- If there is any new rules of abaqus input file format, plz let us know.
- I tried to transfrom `ABAQUS Input file with special format` to `exodus`, the `FileMeshGenerator` is `PASS`.

kind regards,
tianwen

**Comment by lindsayad** [comment]
@roystgnr were there any recent changes to the abaqus reader?

**Reply by roystgnr** [reply]
https://github.com/libMesh/libmesh/pull/3640 just made it into the submodule.

We clearly need more test coverage here.  Could @Ttw0626 post a mesh that was importing before but is failing now?

**Reply by Ttw0626** [reply]
yeah, thanks for reply, @roystgnr @lindsayad. 
**I post two different important and needed format of abaqus input file which were both PASS before.**

`MOOSE input file`: **test_abq_input.i**

`test1 abaqus input file`: **tensile_sample_test1.inp**

`test2 abaqus input file`: **poly_sample_test2.inp**

`previous successful exodus mesh file by "mesh-only"`: **/previous_suc_result_By_FileMeshGenerator/poly_sample_test2.e**

[post_file_abaqus.zip](https://github.com/idaholab/moose/files/13302605/post_file_abaqus.zip)


kind regards,
tianwen

**Reply by roystgnr** [reply]
I can verify the failures with libMesh alone; I'm looking into it there now.

**Reply by roystgnr** [reply]
And I've found one obvious bug.  After fixing that, both inputs read without throwing exceptions.  That doesn't necessarily mean they read *correctly*, though.  Would you give https://github.com/libMesh/libmesh/pull/3699 a try, if you have any regression testing to hit it with?

Also, would it be okay for us to ~~steal~~ copy those meshes to add to the libMesh unit test suite, and if not then could you provide some meshes which exhibit the problem that we could use?  We're in dire need of more test coverage to defend against future regressions.

**Comment by Ttw0626** [comment]
@roystgnr Thanks for this work

**(1)**
I could not understand what should I do in `Would you give https://github.com/libMesh/libmesh/pull/3699 a try, if you have any regression testing to hit it with?`. could you explain more about what I should supply to u ?

**(2)**
By the way, I followed your modification of `abaqus_io.C` in libmesh, it was work, however as you said before, this doesn't mean 
the libmesh read correctly. As a temperary option, I followed your modification.

**(3)**
For me, Plz feel free to use these above abaqus input files which I posted.

Even though these files generated by myself, should we cite the ABAQUS software due to the commercial one ?

kind regards,
tianwen

**Reply by roystgnr** [reply, marked as answer]
"give https://github.com/libMesh/libmesh/pull/3699 a try" == that pull request is for a branch of libMesh, which at that time had an extra commit to it with a proposed fix; I was hoping you could try building that commit into your own codes to see what happens.

"if you have any regression testing to hit it with" == since the testing done automatically on libMesh PRs clearly had inadequate coverage of Abaqus reads (otherwise we'd never have let this bug slip through), but you're doing more complete work with Abaqus files (otherwise the bug wouldn't have hit you), I'm hoping you might have your own test suite for that work, which might include a "regression test" - a problem with a 'gold' expected solution (from an analytic solution or manufactured solution or at least from previous runs) and a test framework which checks new code versions' solution against that expectation.  Even if you didn't have to worry about library regressions (for which I do apologize) this would be an important thing to have to check for bugs at higher levels.

I'll add your files to the unit test suite (in additional commits to https://github.com/libMesh/libmesh/pull/3699) so we can at least avoid regressions this gross in the future; thanks very much for the contribution.  I'll mention that they were ABAQUS-generated in the comments there.

And I might be able to come up with something to get better testing than my first crack of "does this load; do the node and element counts match".  If I converted these files to another format and then added a test for equality of reads, that would cover all the subdomain/sideset type information that a simple test would miss...

**Reply by Ttw0626** [reply]
Oh, I see, you gave a super-nice explanation, I did not know before.
I noticed your newest merge request #https://github.com/libMesh/libmesh/pull/3699.

It is impossible for me to learn all of libmesh source code, Thanks for your help agian,
If you has any thing which I could supply for u, plz feel free to ask.

If the libmesh submodule of moose (master branch) is updated in the future,
this issue should be closed.

kind regards,
tianwen
